src/musicPlaylistSecondary.java

```diff
@@ -0,0 +1,409 @@
+import java.util.Random;
+
+public abstract class musicPlaylistSecondary implements musicPlaylist {
+
+    /**
+     * Private helper: reports whether this playlist is empty by checking the
+     * kernel method getCurrentSong().
+     *
+     * @return true if the playlist is empty, false otherwise
+     */
+    private boolean isEmpty() {
+        // If getCurrentSong() returns the default "No songs in playlist",
+        // we consider it empty.
+        return "No songs in playlist".equals(this.getCurrentSong());
+    }
+
+    /*
+     * ------------------------------------------------------------------------
+     * KERNEL METHODS IMPLEMENTATION (from MusicPlaylistKernel)
+     * ------------------------------------------------------------------------
+     */
+
+    @Override
+    public void addSong(String song) {
+
+        if (song == null) {
+            // If we must handle null, do nothing or throw an exception
+            return;
+        }
+        musicPlaylist temp = this.newInstance();
+        temp.transferFrom(this); // now 'this' is empty, and 'temp' has the songs
+
+        // Re-transfer from temp to this, preserving order
+        if (!temp.isEmpty()) {
+            // Mark the first song
+            String start = temp.getCurrentSong();
+            // Move the first song from temp->this
+            this.addSongKernel(temp, temp.getCurrentSong());
+            temp.removeSong(start);
+
+            // Rotate temp until we are back at 'start'
+            while (!temp.isEmpty() && !temp.isEmptyPlaylistWithSong(start)) {
+                this.addSongKernel(temp, temp.getCurrentSong());
+                temp.removeSong(temp.getCurrentSong());
+            }
+        }
+
+        // Now 'this' has the old songs in the same order, 'temp' is empty
+        //  Finally, add the new song
+        // We "append" by just calling the kernel approach to add a new one
+        this.addSongKernel(null, song);
+    }
+
+    /**
+     * A helper to add a song from "source" to "this" while simulating
+     * "preserving order". Essentially, we rely on kernel-level approach but do
+     * not break the layering.
+     *
+     * @param source
+     *            optional playlist providing context (can be null)
+     * @param song
+     *            the song to add
+     */
+    private void addSongKernel(musicPlaylist source, String song) {
+        this.superAddSong(song);
+    }
+
+    /**
+     * A minimal version of adding a song without all the shuffle/transfer
+     * logic, in practice calling super. This is to illustrate "kernel-level"
+     * calls in this contrived example.
+     *
+     * @param song
+     *            the song to add
+     */
+    private void superAddSong(String song) {
```

Comment:
Reminder that contracts aren't needed here. They're inherited. You should probably put `@override` above the methods instead.